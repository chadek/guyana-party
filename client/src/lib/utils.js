import axios from 'axios'
import Cookies from 'js-cookie'
import dompurify from 'dompurify'
import Showdown from 'showdown'
import md5 from 'md5'
import cogoToast from 'cogo-toast'

export const MISSING_TOKEN_ERR = 'Token de connexion requis'

export const reload = () => {
  if (typeof window !== 'undefined') window.location.reload()
}

export const gravatar = email => `https://www.gravatar.com/avatar/${md5(email)}?d=retro`

export const purify = async dirty => dompurify.sanitize(dirty)

export const markToSafeHTML = markdown => {
  const converter = new Showdown.Converter({
    tables: true,
    simplifiedAutoLink: true,
    strikethrough: true,
    tasklists: true
  })
  return purify(converter.makeHtml(markdown))
}

export const getUID = () => Cookies.get('gp_uid')

export const getToken = () => ({ jwt: Cookies.get('gp_jwt'), uid: getUID() })

export const fetcher = url => axios.get(url).then(r => r.data)

export const axiosGet = (url, next, fallback) =>
  axios
    .get(url)
    .then(payload => next(payload))
    .catch(fallback)

export const axiosPost = ({ url, data }, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.post(url, data, options).then(next).catch(fallback)
}

export const axiosPut = ({ url, data }, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.put(url, data, options).then(next).catch(fallback)
}

export const axiosDelete = (url, next, fallback) => {
  const jwt = Cookies.get('gp_jwt')
  if (!jwt) return fallback(MISSING_TOKEN_ERR)

  const options = { headers: { authorization: `bearer ${jwt}` } }

  return axios.delete(url, options).then(next).catch(fallback)
}

/* Convert the coordinate (lat or lng) to DMS (degrees minutes seconds). */
const toDMS = coord => {
  const absolute = Math.abs(coord)
  const degrees = Math.floor(absolute)
  const minutesNotTruncated = (absolute - degrees) * 60
  const minutes = Math.floor(minutesNotTruncated)
  const seconds = Math.floor((minutesNotTruncated - minutes) * 60)
  return `${degrees}° ${minutes}′ ${seconds}″`
}

export const gpsCoords = (lat, lng) => {
  const northSouth = `${toDMS(lat)} ${Math.sign(lat) >= 0 ? 'N' : 'S'}`
  const eastWest = `${toDMS(lng)} ${Math.sign(lng) >= 0 ? 'E' : 'W'}`
  return `${northSouth}, ${eastWest}`
}

export const compress = (files, next) => {
  if (typeof window === 'undefined') return // window not defined in ssr
  const Compress = require('client-compress') // eslint-disable-line global-require
  const compressor = new Compress({ targetSize: 1.0, quality: 0.75 })
  compressor.compress(files).then(data => next(data))
}

export const scrollTo = selector => {
  const target = document.querySelector(selector)
  if (target) target.scrollIntoView({ behavior: 'smooth', block: 'end' })
}

export const toast = (msg, variant = 'info', heading = '') => {
  const { hide } = cogoToast[variant](msg, {
    hideAfter: 5,
    heading: heading || (variant === 'error' ? 'Erreur !' : ''),
    onClick() {
      hide()
    }
  })
}
