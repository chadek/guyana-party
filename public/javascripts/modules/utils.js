import axios from "axios";
import dompurify from "dompurify";

export function axiosGet(url, callback) {
  axios
    .get(url)
    .then(res => callback(res.data))
    .catch(err => console.error(err));
}

export function data2HTML(data, formatFn, concatFormat) {
  if (!data.items) return "";
  const rawHTML = data.items
    .map(item => formatFn(item))
    .join("")
    .concat(concatFormat);
  return dompurify.sanitize(rawHTML);
}

export function formatDateTime(isoDate, fr = true, separator = '/') {
  if (!isoDate) return;
  isoDate = new Date(isoDate);
  let time = ("0" + isoDate.getHours()).slice(-2) + ":" + ("0" + isoDate.getMinutes()).slice(-2);
  let day = ("0" + isoDate.getDate()).slice(-2);
  let month = ("0" + (isoDate.getMonth() + 1)).slice(-2);
  let date = "";
  if(fr) {
    date = `${day}${separator}${month}${separator}${isoDate.getFullYear()}`;
  } else {
    date = `${isoDate.getFullYear()}${separator}${month}${separator}${day}`;
  }
  return {time, date};
}

export function formatTime() {}
