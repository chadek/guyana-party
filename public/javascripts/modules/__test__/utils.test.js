import axios from 'axios'
import { addNewBtn, axiosGet, pagination, sliceStr } from '../utils'

describe('Utils: addNewBtn', () => {
  it('renders the add button correctly', () => {
    expect(addNewBtn()).toMatchSnapshot()
    expect(addNewBtn(true)).toMatchSnapshot()
  })
})

jest.mock('axios')

describe('Utils: axiosGet', () => {
  beforeEach(() => {
    axios.mockClear()
  })

  it('should return results', () => {
    axios.get.mockResolvedValue({ data: { name: 'Chris' } })
    axiosGet('fake-url', d => expect(d.name).toBe('Chris'))
  })

  it('should fail', () => {
    axios.get.mockRejectedValue({ err: 'bad' })
    axiosGet('fake-url', d => expect(d.err).toBe('bad'))
  })
})

describe('Utils: pagination', () => {
  it('displays first page with next button', () => {
    expect(pagination(1, 3)).toMatchSnapshot()
  })

  it('displays second page with previous and next buttons', () => {
    expect(pagination(2, 3)).toMatchSnapshot()
  })

  it('displays third page with previous button', () => {
    expect(pagination(3, 3)).toMatchSnapshot()
  })
})

describe('Utils: sliceStr', () => {
  it('works without limit parameter', () => {
    expect(sliceStr('Un test de chaîne coupée')).toBe(
      'Un test de chaîne co ...'
    )
  })

  it('works without limit parameter, short string', () => {
    expect(sliceStr('Un')).toBe('Un')
  })

  it('works with limit parameter', () => {
    expect(sliceStr('Un test de chaîne coupée', 10)).toBe('Un test de ...')
  })
})
