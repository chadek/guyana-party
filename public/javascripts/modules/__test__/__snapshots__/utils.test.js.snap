// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils: addNewBtn renders the add button correctly 1`] = `
"<div>
    <div class=\\"card card__new card__new--event\\" title=\\"Ajouter un évènement\\"></div>
  </div>"
`;

exports[`Utils: addNewBtn renders the add button correctly 2`] = `
"<div>
    <div class=\\"card card__new card__new--group\\" title=\\"Ajouter un groupe\\"></div>
  </div>"
`;

exports[`Utils: pagination displays first page with next button 1`] = `"<div class=\\"pagination text-center\\"><ul><li class=\\"active\\" title=\\"page 1\\">1</li><li class=\\"pageBtn\\" title=\\"page 2\\">2</li><li class=\\"pageBtn\\" title=\\"page 3\\">3</li><li class=\\"pageBtn\\" title=\\"page suivante\\">&raquo;</li></ul></div>"`;

exports[`Utils: pagination displays second page with previous and next buttons 1`] = `"<div class=\\"pagination text-center\\"><ul><li class=\\"pageBtn\\" title=\\"page précédente\\">&laquo;</li><li class=\\"pageBtn\\" title=\\"page 1\\">1</li><li class=\\"active\\" title=\\"page 2\\">2</li><li class=\\"pageBtn\\" title=\\"page 3\\">3</li><li class=\\"pageBtn\\" title=\\"page suivante\\">&raquo;</li></ul></div>"`;

exports[`Utils: pagination displays third page with previous button 1`] = `"<div class=\\"pagination text-center\\"><ul><li class=\\"pageBtn\\" title=\\"page précédente\\">&laquo;</li><li class=\\"pageBtn\\" title=\\"page 1\\">1</li><li class=\\"pageBtn\\" title=\\"page 2\\">2</li><li class=\\"active\\" title=\\"page 3\\">3</li></ul></div>"`;
